### ISSUE TITLE
Destroy ASG on timeout

### ISSUE THREAD

**Issue by chiefy** [hashicorp/terraform#2381]
Right now I am running into a problem where I am doing rolling deploys by creating new ASG / LCs w/ create-before-destroy. If the ASG health check times out, Terraform doesn't really do anything aside from providing feedback to that fact (which is fine). The issue is that if there is already a healthy deploy (ASG1), now there is an ASG2 out there which is infinitely spawning / terminating instances because there is some kind of problem. 

The other issue is: we have a deploy problem, but hey, at least there is a healthy ASG(ASG1) out there, along with the new bad one (ASG2). If we create another deploy / ASG which is successful, Terraform cleans up the bad one(ASG2) but ASG1 is orphaned. I am not sure if there's any way around this with how TF manages state, other than cleaning up failed ASGs using create-before-destroy

**tl;dr** It'd be nice if there was an option in AWS ASG to destroy on timeout / failure.

**Comment by phinze** [comment]
Thanks for the report - I've been in this situation myself, and I agree that there's more Terraform can do to handle it.

From my understanding of the way Terraform behaves in these scenarios, I don't believe ASG1 should ever be orphaned. What I'd expect to happen is:

First run:

 * ASG1 up and healthy
 * ASG2 created, cannot launch instances, times out
 * ASG1 untouched, ASG2 marked "tainted"

Second run (after correcting config):

 * Tainted ASG2 is destroyed
 * ASG2' is created, successfully launches instances to capacity
 * ASG1 is now destroyed, having been replaced by ASG2

If that's not the behavior, then I'd consider this a bug. Either way, the "destroy on timeout/failure" option is interesting. Will ponder that a bit.

**Comment by phinze** [comment]
Coming back around to this, Terraform's behavior here is as I described it in the last comment. Effectively this _is_ Terraform destroying the timed out / failed ASG, it just does so on the subsequent run. Feel free to follow up if you have further questions on this. :+1:
